homeassistant/components/sensor/gtfs.py

```diff
@@ -228,71 +301,201 @@ def unit_of_measurement(self):
+                    _LOGGER.warning(
+                        "Agency ID '%s' not found in agency table. You may "
+                        "want to update the agency database table to fix this "
+                        "missing reference.", self._route.agency_id)
+                    self._agency = False
+
+            self.update_attributes()
 
             name = '{} {} to {} next departure'
             self._name = (self._custom_name or
-                          name.format(agency.agency_name,
-                                      origin_station.stop_id,
-                                      destination_station.stop_id))
-
-            # Build attributes
-            self._attributes = {}
-            self._attributes['offset'] = self._offset.seconds / 60
-
-            self._icon = ICONS.get(route.route_type, ICON)
-
-            def dict_for_table(resource):
-                """Return a dict for the SQLAlchemy resource given."""
-                return dict((col, getattr(resource, col))
-                            for col in resource.__table__.columns.keys())
-
-            def append_keys(resource, prefix=None):
-                """Properly format key val pairs to append to attributes."""
-                for key, val in resource.items():
-                    if val == "" or val is None or key == 'feed_id':
-                        continue
-                    pretty_key = key.replace('_', ' ')
-                    pretty_key = pretty_key.title()
-                    pretty_key = pretty_key.replace('Id', 'ID')
-                    pretty_key = pretty_key.replace('Url', 'URL')
-                    if prefix is not None and \
-                       pretty_key.startswith(prefix) is False:
-                        pretty_key = '{} {}'.format(prefix, pretty_key)
-                    self._attributes[pretty_key] = val
-
-            append_keys(dict_for_table(agency), 'Agency')
-            append_keys(dict_for_table(route), 'Route')
-            append_keys(dict_for_table(trip), 'Trip')
-            append_keys(dict_for_table(origin_station), 'Origin Station')
-            append_keys(dict_for_table(destination_station),
-                        'Destination Station')
-            append_keys(origin_stop_time, 'Origin Stop')
-            append_keys(destination_stop_time, 'Destination Stop')
+                          name.format(getattr(self._agency, 'agency_name',
+                                              DEFAULT_NAME),
+                                      self.origin,
+                                      self.destination))
+
+    def update_attributes(self):
+        """Update state attributes."""
+        def dict_for_table(resource) -> dict:
+            """Return a dict for the SQLAlchemy resource given."""
+            return dict((col, getattr(resource, col))
+                        for col in resource.__table__.columns.keys())
+
+        def append_keys(resource, prefix=None) -> None:
```

Comment:
Move out of this function definition. Also, do not use pretty keys but instead use key lowercase and underscores. The backend should not be aware of the frontend.

Reply:
Will do! However this will introduce a breaking change as pretty keys have been in use with this component since its introduction.

Reply:
Fixed! I'll add a comment regarding the breaking change in the PR description.